src/types/pointer.md

```diff
@@ -50,6 +50,16 @@ Raw pointers can be created directly using [`core::ptr::addr_of!`] for `*const`
 
 The standard library contains additional 'smart pointer' types beyond references and raw pointers.
 
+## Bit validity
+
+Despite pointers and references being similar to `usize`s in the machine code emitted on most platforms,
+the semantics of transmuting a reference or pointer type to a non-pointer type is currently undecided.
+Thus, it may not be valid to transmute a pointer or reference type, `P`, to a `[u8; size_of::<P>()]`.
```

Comment:
@RalfJung For my own edification, how does this interact with pointer-to-integer `as` casts? I can imagine a few possibilities (not mutually exclusive):
- `as` casts are sound, but transmutations are not necessarily sound (even though you can emulate a transmutation by doing an `as` cast followed by a transmutation from `usize`)
- the first bullet is true _and_ pointer-to-integer `as` casts probably shouldn't have been allowed in safe Rust, but it's too late

Reply:
The current working model is that `as` casts are special:
- `ptr as usize` is like https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.expose_addr
- `usize as ptr` is like https://doc.rust-lang.org/nightly/std/ptr/fn.from_exposed_addr.html

In contrast:
- "transmute ptr to usize" is like https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.addr
- "transmute usize to ptr" is like https://doc.rust-lang.org/nightly/std/ptr/fn.invalid.html

All of these operations are safe and sound, but as their docs show, they behave quite differently and that can affect the soundness of later code that uses the values they produce.

Note that this is a prototype model, which has some unresolved problems and which is very much not stabilized.

Reply:
Okay that all makes sense, thanks!

> Note that this is a prototype model, which has some unresolved problems and which is very much not stabilized.

I always assume that of everything in this space unless documented otherwise :)